Add docs for api conventions and analyzers (#9453)
aspnet/Docs: aspnetcore/web-api/advanced/api-conventions.md

---
title: Using web API conventions
author: pranavkm
description: Learn about web API conventions

> ms.author: pranavkm

Reviewer: Add the following metadata above this line, so that the doc is only available for ASP.NET Core 2.2+:

```
monikerRange: '>= aspnetcore-2.2'
```

ms.custom: mvc

> ms.date: 11/5/2018

Reviewer: ```suggestion
ms.date: 11/05/2018
```

uid: web-api/api-conventions
---
# Learn about web API conventions

> ASP.NET Core 2.2 introduces a way to extract common [API documentation](<xref:tutorials/web-api-help-pages-using-swagger>) and apply it multiple actions, or controllers, or all controllers within an assembly. Web API conventions are a substitute to decorating individual actions with <xref:Microsoft.AspNetCore.Mvc.ProducesResponseTypeAttribute>. It let you define the most common "conventional" return types and status codes that you return from your action with a way to select the convention method that applies to an action.

Reviewer: ```suggestion
ASP.NET Core 2.2 introduces a way to extract common [API documentation](xref:tutorials/web-api-help-pages-using-swagger) and apply it to multiple actions, controllers, or all controllers within an assembly. Web API conventions are a substitute for decorating individual actions with <xref:Microsoft.AspNetCore.Mvc.ProducesResponseTypeAttribute>. It allows you to define the most common "conventional" return types and status codes that you return from your action with a way to select the convention method that applies to an action.
```

By default, ASP.NET Core MVC 2.2 ships with a set of default conventions, <xref:Microsoft.AspNetCore.Mvc.DefaultApiConventions>, that is based on the controller that ASP.NET Core scaffolds. If your actions follow the pattern that scaffolding produces, you should be successful using the default conventions.

At runtime, ApiExplorer understand conventions. ApiExplorer is MVC’s abstraction to communicate with Open API document generators. Attributes from the applied convention get associated with an action and will be included in action’s Swagger documentation. API analyzers also understand conventions. If your action is unconventional i.e. it returns a status code that is not documented by the applied convention, it will produce a warning, encouraging you to document it.

> # Applying web API conventions

Reviewer: ```suggestion
## Apply web API conventions
```

1. <xref:Microsoft.AspNetCore.Mvc.ApiConventionMethodAttribute> - Applies to individual actions and specifies the convention type and the convention method that applies. In the sample below, the convention method <xref:Microsoft.AspNetCore.Mvc.DefaultApiConventions.Put> is applied to the `Update` action:

[!code-csharp[](api-conventions/sample/Controllers/ContactsConventionController.cs?name=apiconventionmethod&highlight=2-3)]

2. <xref:Microsoft.AspNetCore.Mvc.ApiConventionTypeAttribute> applied to a controller - Applies the convention type to all actions on the controller. Convention methods are decoarated with hints that determine which actions it would apply to (details as part of authoring conventions).

[!code-csharp[](api-conventions/sample/Controllers/ContactsConventionController.cs?name=apiconventiontypeattribute)]

> 3.. <xref:Microsoft.AspNetCore.Mvc.ApiConventionTypeAttribute> applied to an assembly - Applies the convention type to all controllers in the current assembly. e.g.

Reviewer: e.g. --> For example:

[!code-csharp[](api-conventions/sample/Startup.cs?name=apiconventiontypeattribute)]

> # Authoring web API conventions

Reviewer: ```suggestion
## Create web API conventions
```

A convention is a static type with methods. These methods are annotated with <xref:Microsoft.AspNetCore.Mvc.ProducesResponseTypeAttribute> or <xref:Microsoft.AspNetCore.Mvc.ProducesDefaultResponseTypeAttribute> attributes.

> ```C#

Reviewer: ```suggestion
```csharp
```

public static class MyAppConventions
{
    [ProducesResponseType(200)]
    [ProducesResponseType(404)]
    public static void Find(int id)
    {

    }
}
```

Applying this convention to an assembly would result in the convention method applying to any action with the name Find and having exactly one parameter named id, as long as they do not have other more specific metadata attributes.

In addition to ProducesResponseType and ProducesDefaultResponseType, two additional attributes – <xref:Microsoft.AspNetCore.Mvc.ApiExplorer.ApiConventionNameMatchAttribute> and <xref:Microsoft.AspNetCore.Mvc.ApiExplorer.ApiConventionTypeMatchAttribute> – can be applied to the convention method that determines the methods they apply to.

> ```C#

Reviewer: ```suggestion
```csharp
```

[ProducesResponseType(200)]
[ProducesResponseType(404)]
[ApiConventionNameMatch(ApiConventionNameMatchBehavior.Prefix)]
public static void Find(
    [ApiConventionNameMatch(ApiConventionNameMatchBehavior.Suffix)]
    int id)
{ }
```

> * The <xref:Microsoft.AspNetCore.Mvc.ApiExplorer.ApiConventionNameMatchBehavior.Prefix> option applied to the method, indicates that the convention can match any action as long as it starts with the prefix “Find”. This will include methods such as Find, FindPet or FindById.

Reviewer: Add a comma after "FindPet".
